### PR TITLE
Add Segment Routing IS-IS BGP-LS extensions

### PR DESCRIPTION
**Pull request by ruissalo**
@thomas-mangin I'm creating a PR so you can take a look before merging. It shouldn't break anything. I also added a packet hex dump to the qa/tests for these changes.

This PR contains

-  IS-IS segment routing extensions to BGP-LS (draft-gredler-idr-bgp-ls-segment-routing-ext-03).
Added TLVs mostly correspond to those currently implemented by IOS-XR 6.1.1.  Binding SIDs extensions are not included but I don't think Cisco supports them at the moment.

Use Case: Segment IDs are learned via ExaBGP/BGP-LS (with SR extensions enabled) => controller application computes the constrained LSP using the segment IDs previously learned => LSP is programmed via ExaBGP/BGP-LU session.

I'll update the wiki with a blog post on how ExaBGP can be used as an SDN controller to program LSPs using segment routing instructions.

- Removes duplicated code in several TLVs:  Flags parsing is moved to a separate class.

---
This change is [Reviewable](https://reviewable.io/reviews/exa-networks/exabgp/615)

### CODE REVIEW

File: lib/exabgp/bgp/message/update/attribute/bgpls/linkstate.py [hunk 1]
```diff
@@ -72,11 +75,52 @@ def __init__ (self, code, rep):
...
+		self.flags = flags
+
+	@classmethod
+	def unpack(cls,data,pattern):
+		pad = pattern.count('RSV')
+		repeat = len(pattern) - pad
+		flag_array = binascii.b2a_hex(data)
+		hex_rep = hex(int(flag_array, 16))
+		bit_array = BitArray(hex_rep)
+		valid_flags = [''.join(item) + '0' * pad for item in itertools.product('01', repeat=repeat)]
```
Comment (thomas-mangin) [inline lib/exabgp/bgp/message/update/attribute/bgpls/linkstate.py]: This is not python 3 friendly .. I think we have a function to perform string concatenation

**Comment by thomas-mangin** [comment]
looks very good - just need to make sure we use the helper functions which are required for python3

File: lib/exabgp/bgp/message/update/attribute/bgpls/linkstate.py (hunk 1 above)
Comment (thomas-mangin) [inline lib/exabgp/bgp/message/update/attribute/bgpls/linkstate.py]: from exabgp.util import concat_strs

**Comment by thomas-mangin** [comment]
Sorry my latest work to remove all pylint errors caused a conflict - very sorry

**Comment by ruissalo** [comment]
Hi @thomas-mangin , no worries, I'll fix the conflicts and use concat_strs tomorrow.

**Comment by thomas-mangin** [comment]
Going to release 4.0 very soon now .. would be great to have it before I do 😉

**Comment by ruissalo** [comment]
great news! I'll merge tomorrow.

**Comment by thomas-mangin** [comment]
👍

**Pull request merged**